src/compiler/scala/tools/nsc/classpath/DirectoryClassPath.scala

```diff
@@ -113,9 +130,23 @@ trait JFileDirectoryLookup[FileEntryType <: ClassRepresentation] extends Directo
     util.Arrays.sort(listing, (o1: File, o2: File) => o1.getName.compareTo(o2.getName))
     listing
   }
+
+  override protected def foreachChild(dir: File)(f: (File, BasicFileAttributes) => Unit): Unit = {
+    import java.nio.file._
+    val result = new collection.mutable.ArrayBuffer[Path]()
+    val visitor = new SimpleFileVisitor[Path]() {
+      override def visitFile(subdir: Path, attrs: attribute.BasicFileAttributes): FileVisitResult = {
+        if (dir != subdir) f(subdir.toFile, attrs)
+        FileVisitResult.CONTINUE
+      }
+    }
+    // OPT: Prefer walkFileTree to File#listFiles, especially on Windows: https://bugs.openjdk.java.net/browse/JDK-8008469
+    Files.walkFileTree(dir.toPath, util.EnumSet.of(FileVisitOption.FOLLOW_LINKS), 1, visitor)
```

Comment:
Hmm, we probably will to keep a temporary collection after all, so that we can still sort by file name [as before](https://github.com/scala/scala/pull/7140/commits/f611e0fd04021d884b448de85e5a5bf425964976#diff-58a86ecff0e6f9ae84d4ddea018e5eb4R130)

Reply:
We could defer the sorting to the callsites where necessary

Reply:
We're using `collection.Seq` in some places alread (`ClassPathEntries`) to allow mutable collections. Maybe we can extend that (and use in-place sorting)?